Documentation/coding-guidelines/interop-pinvokes.md

```diff
@@ -0,0 +1,240 @@
+P/Invokes
+=========
+
+This document extends the [Interop Guidelines](interop-guidelines.md) to provide more specific guidelines, notes, and resources for defining P/Invokes.
+
+Guidelines
+----------
+1. Avoid StringBuilder
+2. Use `ExactSpelling` where possible
+3. Don't use `SetLastError` where not appropriate
+4. Be careful of `BOOL` vs `BOOLEAN`
+5. Watch for buffer size for output strings (null/no null)
+
+Attributes
+----------
+
+**Implicit attributes applied to parameter and return values:**
+
+|                  | Implicit Attribute |
+|------------------|--------------------|
+| parameter        | `[In]`             |
+| `out` parameter  | `[Out]`            |
+| `ref` parameter  | `[In],[Out]`       |
+| return value     | `[Out]`            |
+
+**`[DllImport()]` [1] attribute settings:**
+
+| Setting | Recommendation | Details |
+|---------|----------------|---------|
+| [`PreserveSig`][2]   | keep default       | When this is explicitly set to false (the default is true), failed HRESULT return values will be turned into Exceptions (and the return value in the definition becomes null as a result).|
+| [`SetLastError`][3]  | as per API         | Set this to true (default is false) if the API uses GetLastError and use Marshal.GetLastWin32Error to get the value. If the API sets a condition that says it has an error, get the error before making other calls to avoid inadvertently having it overwritten.|
+| [`ExactSpelling`][4] | `true`             | Set this to true (deafult is false) and gain a slight perf benefit as the framework will avoid looking for an "A" or "W" version. (See NDirectMethodDesc::FindEntryPoint).|
+| [`CharSet`][5]       | `CharSet.Unicode` if strings present in signature | This specifies marshalling behavior of strings and what `ExactSpelling` does when `false`. Be explicit with this one as the documented default is `CharSet.Ansi`.|
```

Comment:
Ansi (which is UTF8 on Unix) is almost always the right answer.  We're supposed to be getting a CharSet.Utf8 (we may already have it merged, haven't checked), at which point that'd almost always be the right answer.